### PR TITLE
Reinfer discard in assignment and out vars

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/36503 (warning in discarding assignment)
Fixes https://github.com/dotnet/roslyn/issues/29635 (reinfering discards), but issue https://github.com/dotnet/roslyn/issues/35010 tracks the remaining scenarios around deconstructions (including discards there)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -5402,7 +5417,8 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
                 CheckDisallowedNullAssignment(rightState, leftAnnotations, right.Syntax.Location);
 
                 TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightState, MakeSlot(right));
-                SetResult(node, TypeWithState.Create(leftLValueType.Type, rightState.State), leftLValueType);
+                var resultState = left is BoundDiscardExpression ? rightState : TypeWithState.Create(leftLValueType.Type, rightState.State);
+                SetResult(node, resultState, leftLValueType);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >leftLValueType [](start = 45, length = 14)

Is this correct for the discard?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -5391,7 +5400,13 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
...
                 if (!node.IsRef)
                 {
-                    rightState = VisitOptionalImplicitConversion(right, leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
+                    var discarded = left is BoundDiscardExpression;
+                    rightState = VisitOptionalImplicitConversion(right, targetTypeOpt: discarded ? default : leftLValueType, UseLegacyWarnings(left, leftLValueType), trackMembers: true, AssignmentKind.Assignment);
+
+                    if (left is BoundDiscardExpression discard)
+                    {
+                        SetAnalyzedNullability(discard, new VisitResult(rightState), isLvalue: true);
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Is this different from the `SetResult` below?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: One is for the discard `_` and the other is for the assignment expression `_ = expr`.

---
In reply to: [303490371](https://github.com/dotnet/roslyn/pull/37214#discussion_r303490371) [](ancestors = 303490371)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -44262,38 +44262,153 @@ static void F((object, object?) t)
...
+        {
+            var source = @"
+#nullable enable
+interface I<in T> {}
+
+class C
+{
+    void M(I<I<string?>> i1, I<I<string>?> i2)
+    {
+        var x = i2 ?? i1;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I'm struggling to understand why this works. It seems like the innermost `string` -> `string?` conversion is not safe?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -44262,38 +44262,153 @@ static void F((object, object?) t)
...
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+            var arguments = tree.GetRoot().DescendantNodes().OfType<ArgumentSyntax>();
+
+            var discard1 = arguments.First().Expression;
+            Assert.Equal("var _", discard1.ToString());
+            Assert.Equal("System.String", model.GetTypeInfo(discard1).Type.ToTestDisplayString());
+            Assert.Equal(CodeAnalysis.NullableAnnotation.Annotated, model.GetTypeInfo(discard1).Nullability.Annotation);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: are there test cases already existing which exercise the annotation on a non-nullable discard?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I have to say I have trouble with complex cases like this too. The important thing which I wanted to verify here is that `_` works like `var` declaration.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Added

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 5]
```diff
@@ -44262,38 +44262,166 @@ static void F((object, object?) t)
...
+
+            var discard2 = arguments.Skip(1).First().Expression;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >arguments.Skip(1).First() [](start = 27, length = 25)

Minor point: `arguments.ElementAt(1)`

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Thanks. I'll remember for next time. I'll leave as-is to meet today's deadline though.

**Pull request merged**